src/libstd/rt.rs

```diff
@@ -26,7 +26,58 @@
 // Reexport some of our utilities which are expected by other crates.
 pub use panicking::{begin_panic, begin_panic_fmt, update_panic_count};
 
-#[cfg(not(test))]
+// To reduce the generated code of the new `lang_start`, this function is doing
+// the real work.
+#[cfg(not(any(test, stage0)))]
+fn lang_start_internal(main: &(Fn() -> i32 + Sync + ::panic::RefUnwindSafe),
+                       argc: isize, argv: *const *const u8) -> ! {
+    use panic;
+    use sys;
+    use sys_common;
+    use sys_common::thread_info;
+    use thread::Thread;
+    use process;
+    #[cfg(not(feature = "backtrace"))]
+    use mem;
+
+    sys::init();
+
+    process::exit(unsafe {
```

Comment:
From our discussion on gitter, this is the cause of the problem in wasm (usage of `process::exit` which is [an abort](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/wasm/os.rs#L130-L132) on wasm).

Let's change this to propagate the error code outwards I think?